samza-core/src/main/scala/org/apache/samza/util/CommandLine.scala

```diff
@@ -19,55 +19,60 @@
 
 package org.apache.samza.util
 
-import java.net.URI
-import joptsimple.{OptionParser, OptionSet}
+import joptsimple.{ArgumentAcceptingOptionSpec, OptionParser, OptionSet}
 import joptsimple.util.KeyValuePair
-import org.apache.samza.config.{ConfigFactory, MapConfig}
-import org.apache.samza.config.factories.PropertiesConfigFactory
-import scala.collection.mutable.Buffer
+import org.apache.samza.config.{Config, ConfigLoaderFactory, JobConfig, MapConfig}
+import org.apache.samza.config.loaders.PropertiesConfigLoaderFactory
+
 import scala.collection.JavaConverters._
+import scala.collection.mutable
 
 /**
  * Defines a basic set of command-line options for Samza tasks. Tools can use this
  * class directly, or subclass it to add their own options.
  */
 class CommandLine {
   val parser = new OptionParser()
-  val configFactoryOpt =
-    parser.accepts("config-factory", "The config factory to use to read your config file.")
+  val configLoaderFactoryOpt: ArgumentAcceptingOptionSpec[String] =
+    parser.accepts("config-loader-factory", "The config factory to use to read your config file.")
           .withRequiredArg
           .ofType(classOf[java.lang.String])
           .describedAs("com.foo.bar.ClassName")
-          .defaultsTo(classOf[PropertiesConfigFactory].getName)
-  val configPathOpt =
-    parser.accepts("config-path", "URI location to a config file (e.g. file:///some/local/path.properties). " +
+          .defaultsTo(classOf[PropertiesConfigLoaderFactory].getName)
+  val configLoaderPropertiesOpt: ArgumentAcceptingOptionSpec[KeyValuePair] =
+    parser.accepts("config-loader-properties", "URI location to a config file (e.g. file:///some/local/path.properties). " +
                                   "If multiple files are given they are all used with later files overriding any values that appear in earlier files.")
           .withRequiredArg
-          .ofType(classOf[URI])
+          .ofType(classOf[KeyValuePair])
           .describedAs("path")
-  val configOverrideOpt =
+  val configOverrideOpt: ArgumentAcceptingOptionSpec[KeyValuePair] =
     parser.accepts("config", "A configuration value in the form key=value. Command line properties override any configuration values given.")
           .withRequiredArg
           .ofType(classOf[KeyValuePair])
           .describedAs("key=value")
 
-  var configFactory: ConfigFactory = null
+  var configLoaderFactory: ConfigLoaderFactory = _
 
-  def loadConfig(options: OptionSet) = {
+  def loadConfig(options: OptionSet): Config = {
     // Verify legitimate parameters.
-    if (!options.has(configPathOpt)) {
+    if (!options.has(configLoaderPropertiesOpt)) {
```

Comment:
Should this check for `configLoaderFactoryOpt` instead? Theoretically, the config loader factory won't need additional properties. You might also need to somehow validate or catch that not all config loader properties were specified for the config loader.

Reply:
configLoaderFactoryOpt has a default value of PropertiesConfigLoaderFactory.

This check is to be consistent with the previous check where we require a config path (URI) to pass to the legacy ConfigFactory, which in theory may not need URI either.

We may remove this check completely too, if we want to support the case where config loader factory does not need any properties at all.

Reply:
In the previous case (i.e. `ConfigFactory`), the `configURI` was a required argument for the interface, and I don't think you can have an empty `URI`. For `ConfigLoaderFactory` interface, it is required to pass a `Config`, but it is possible to have an empty `Config`, so I think you could just remove the check to handle the general case.
The additional complexity that the `ConfigLoaderFactory` brings is that each concrete type requires different configs to be in `Config`, so if there is an error due to a missing config value, then that's maybe where you need to do the `parser.printHelpOn(System.err); System.exit(-1);`.

Reply:
Make sense, removed the check.

The default concrete `PropertiesConfigLoaderFactory` is checking the properties itself too.